# Imports

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [3]:
# path is the raw csv file
path = r"C:\Users\adame\OneDrive\Documents\GitHub\springboard_repository\Capstone_books\Capstone_books\data\raw\books.csv"
df = pd.read_csv(path, on_bad_lines='warn')

Skipping line 3350: expected 12 fields, saw 13
Skipping line 4704: expected 12 fields, saw 13
Skipping line 5879: expected 12 fields, saw 13
Skipping line 8981: expected 12 fields, saw 13



#### Data would not load without an error, required skipping fours lines.

In [4]:
df.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  
0        9/16/2006  Scholastic Inc.  
1         9/1/2004  Scholastic Inc.  
2        11/1/2003       Scholastic  
3         5/1/2004  Scholastic Inc.  
4        9/13/2004       Scholastic

In [5]:
df.tail()

bookID                                             title  \
11118   45631  Expelled from Eden: A William T. Vollmann Reader   
11119   45633                       You Bright and Risen Angels   
11120   45634                   The Ice-Shirt (Seven Dreams #1)   
11121   45639                                       Poor People   
11122   45641                       Las aventuras de Tom Sawyer   

                                                 authors  average_rating  \
11118  William T. Vollmann/Larry McCaffery/Michael He...            4.06   
11119                                William T. Vollmann            4.08   
11120                                William T. Vollmann            3.96   
11121                                William T. Vollmann            3.72   
11122                                         Mark Twain            3.91   

             isbn         isbn13 language_code    num_pages  ratings_count  \
11118  1560254416  9781560254416           eng          512            156   
11119  0140110879  9780140110876           eng          635            783   
11120  0140131965  9780140131963           eng          415            820   
11121  0060878827  9780060878825           eng          434            769   
11122  8497646983  9788497646987           spa          272            113   

       text_reviews_count publication_date      publisher  
11118                  20       12/21/2004  Da Capo Press  
11119                  56        12/1/1988  Penguin Books  
11120                  95         8/1/1993  Penguin Books  
11121                 139        2/27/2007           Ecco  
11122                  12        5/28/2006  Edimat Libros

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11123 non-null  int64  
 1   title               11123 non-null  object 
 2   authors             11123 non-null  object 
 3   average_rating      11123 non-null  float64
 4   isbn                11123 non-null  object 
 5   isbn13              11123 non-null  int64  
 6   language_code       11123 non-null  object 
 7     num_pages         11123 non-null  int64  
 8   ratings_count       11123 non-null  int64  
 9   text_reviews_count  11123 non-null  int64  
 10  publication_date    11123 non-null  object 
 11  publisher           11123 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


In [7]:
df.describe()

bookID  average_rating        isbn13     num_pages  \
count  11123.000000    11123.000000  1.112300e+04  11123.000000   
mean   21310.856963        3.934075  9.759880e+12    336.405556   
std    13094.727252        0.350485  4.429758e+11    241.152626   
min        1.000000        0.000000  8.987060e+09      0.000000   
25%    10277.500000        3.770000  9.780345e+12    192.000000   
50%    20287.000000        3.960000  9.780582e+12    299.000000   
75%    32104.500000        4.140000  9.780872e+12    416.000000   
max    45641.000000        5.000000  9.790008e+12   6576.000000   

       ratings_count  text_reviews_count  
count   1.112300e+04        11123.000000  
mean    1.794285e+04          542.048099  
std     1.124992e+05         2576.619589  
min     0.000000e+00            0.000000  
25%     1.040000e+02            9.000000  
50%     7.450000e+02           47.000000  
75%     5.000500e+03          238.000000  
max     4.597666e+06        94265.000000

In [14]:
df.isnull().sum()

bookID                0
title                 0
authors               0
average_rating        0
isbn                  0
isbn13                0
language_code         0
  num_pages           0
ratings_count         0
text_reviews_count    0
publication_date      0
publisher             0
dtype: int64

In [16]:
df.shape

(11123, 12)

In [17]:
df.dtypes

bookID                  int64
title                  object
authors                object
average_rating        float64
isbn                   object
isbn13                  int64
language_code          object
  num_pages             int64
ratings_count           int64
text_reviews_count      int64
publication_date       object
publisher              object
dtype: object

In [18]:
df.select_dtypes('object')

title  \
0      Harry Potter and the Half-Blood Prince (Harry ...   
1      Harry Potter and the Order of the Phoenix (Har...   
2      Harry Potter and the Chamber of Secrets (Harry...   
3      Harry Potter and the Prisoner of Azkaban (Harr...   
4      Harry Potter Boxed Set  Books 1-5 (Harry Potte...   
...                                                  ...   
11118   Expelled from Eden: A William T. Vollmann Reader   
11119                        You Bright and Risen Angels   
11120                    The Ice-Shirt (Seven Dreams #1)   
11121                                        Poor People   
11122                        Las aventuras de Tom Sawyer   

                                                 authors        isbn  \
0                             J.K. Rowling/Mary GrandPré  0439785960   
1                             J.K. Rowling/Mary GrandPré  0439358078   
2                                           J.K. Rowling  0439554896   
3                             J.K. Rowling/Mary GrandPré  043965548X   
4                             J.K. Rowling/Mary GrandPré  0439682584   
...                                                  ...         ...   
11118  William T. Vollmann/Larry McCaffery/Michael He...  1560254416   
11119                                William T. Vollmann  0140110879   
11120                                William T. Vollmann  0140131965   
11121                                William T. Vollmann  0060878827   
11122                                         Mark Twain  8497646983   

      language_code publication_date        publisher  
0               eng        9/16/2006  Scholastic Inc.  
1               eng         9/1/2004  Scholastic Inc.  
2               eng        11/1/2003       Scholastic  
3               eng         5/1/2004  Scholastic Inc.  
4               eng        9/13/2004       Scholastic  
...             ...              ...              ...  
11118           eng       12/21/2004    Da Capo Press  
11119           eng        12/1/1988    Penguin Books  
11120           eng         8/1/1993    Penguin Books  
11121           eng        2/27/2007             Ecco  
11122           spa        5/28/2006    Edimat Libros  

[11123 rows x 6 columns]

#### The `isbn` column is listed as an object, I have done some research on this. The `isbn` can have a leading zero and for this reason it should remain an object. 

In [19]:
df['isbn']

0        0439785960
1        0439358078
2        0439554896
3        043965548X
4        0439682584
            ...    
11118    1560254416
11119    0140110879
11120    0140131965
11121    0060878827
11122    8497646983
Name: isbn, Length: 11123, dtype: object

#### Check for any letters in the `isbn` column. 

In [21]:
matches = df[df['isbn'].str.contains('[a-zA-Z]', case=False, na=False)]

In [38]:
matches.count()

bookID                984
title                 984
authors               984
average_rating        984
isbn                  984
isbn13                984
language_code         984
  num_pages           984
ratings_count         984
text_reviews_count    984
publication_date      984
publisher             984
dtype: int64

#### There are 984 values in the `isbn` column that contain a letter.

#### Checking for duplicates

In [42]:
df['title'].value_counts()

title
The Iliad                                               9
The Brothers Karamazov                                  9
The Odyssey                                             8
Anna Karenina                                           8
'Salem's Lot                                            8
                                                       ..
The Noonday Demon: An Atlas of Depression               1
The Noonday Demon: An Anatomy of Depression             1
My Secret: A PostSecret Book                            1
The Secret Lives of Men and Women: A PostSecret Book    1
Las aventuras de Tom Sawyer                             1
Name: count, Length: 10348, dtype: int64

In [47]:
# check one of the duplicate titles to see what kind of data is present.
df[df['title'] == 'The Iliad']

bookID      title                                  authors  \
403     1371  The Iliad         Homer/Robert Fagles/Bernard Knox   
405     1374  The Iliad      Homer/Robert Fitzgerald/Andrew Ford   
407     1376  The Iliad  Homer/E.V. Rieu/Peter Jones/D.C.H. Rieu   
408     1377  The Iliad                       Homer/W.H.D. Rouse   
555     1796  The Iliad                  Homer/Robert Fitzgerald   
3355   12254  The Iliad         Homer/Robert Fagles/Derek Jacobi   
5903   22221  The Iliad                                    Homer   
8534   32780  The Iliad                        Homer/Andrew Lang   
8536   32782  The Iliad      Homer/Victor G. Ambrus/Nick McCarty   

      average_rating        isbn         isbn13 language_code    num_pages  \
403             3.86  0140275363  9780140275360           eng          683   
405             3.86  0374529051  9780374529055         en-US          588   
407             3.86  0140447946  9780140447941           eng          462   
408             3.86  0451527372  9780451527370         en-US          312   
555             3.86  1857150600  9781857150605           eng          594   
3355            3.86  0143059289  9780143059288           eng            8   
5903            3.86  0471377589  9780471377580           eng          150   
8534            3.86  1904633382  9781904633389           eng          542   
8536            3.86  0753453215  9780753453216           eng           95   

      ratings_count  text_reviews_count publication_date  \
403          288792                3423        4/29/1999   
405             692                  81         4/3/2004   
407            1919                 118        1/30/2003   
408             158                  15         8/1/1999   
555              30                   2        3/19/1992   
3355             42                   7        6/15/2006   
5903           3834                 134       10/28/1999   
8534             64                   8         9/1/2011   
8536              8                   3        9/15/2000   

                      publisher  
403            Penguin Classics  
405   Farrar  Straus and Giroux  
407            Penguin Classics  
408             Signet Classics  
555                    Everyman  
3355         Penguin-HighBridge  
5903          John Wiley & Sons  
8534        Collector's Library  
8536                 Kingfisher

#### This shows there are nine versions of 'The Iliad' in the data. They all have the same `average_rating`, but everything other column holds a different value. 